vendor/github.com/avast/retry-go/v4/retry.go

```diff
@@ -0,0 +1,313 @@
+
+slightly inspired by [Try::Tiny::Retry](https://metacpan.org/pod/Try::Tiny::Retry)
+
+# SYNOPSIS
+
+http get with retry:
+
+	url := "http://example.com"
+	var body []byte
+
+	err := retry.Do(
+		func() error {
+			resp, err := http.Get(url)
+			if err != nil {
+				return err
+			}
+			defer resp.Body.Close()
+			body, err = ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		},
+	)
+
+	fmt.Println(body)
+
+[next examples](https://github.com/avast/retry-go/tree/master/examples)
+
+# SEE ALSO
+
+* [giantswarm/retry-go](https://github.com/giantswarm/retry-go) - slightly complicated interface.
+
+* [sethgrid/pester](https://github.com/sethgrid/pester) - only http retry for http calls with retries and backoff
+
+* [cenkalti/backoff](https://github.com/cenkalti/backoff) - Go port of the exponential backoff algorithm from Google's HTTP Client Library for Java. Really complicated interface.
+
+* [rafaeljesus/retry-go](https://github.com/rafaeljesus/retry-go) - looks good, slightly similar as this package, don't have 'simple' `Retry` method
+
+* [matryer/try](https://github.com/matryer/try) - very popular package, nonintuitive interface (for me)
+
+# BREAKING CHANGES
+
+* 4.0.0
+  - infinity retry is possible by set `Attempts(0)` by PR [#49](https://github.com/avast/retry-go/pull/49)
+
+* 3.0.0
+  - `DelayTypeFunc` accepts a new parameter `err` - this breaking change affects only your custom Delay Functions. This change allow [make delay functions based on error](examples/delay_based_on_error_test.go).
+
+* 1.0.2 -> 2.0.0
+  - argument of `retry.Delay` is final delay (no multiplication by `retry.Units` anymore)
+  - function `retry.Units` are removed
+  - [more about this breaking change](https://github.com/avast/retry-go/issues/7)
+
+* 0.3.0 -> 1.0.0
+  - `retry.Retry` function are changed to `retry.Do` function
+  - `retry.RetryCustom` (OnRetry) and `retry.RetryCustomWithOpts` functions are now implement via functions produces Options (aka `retry.OnRetry`)
+*/
+package retry
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+)
+
+// Function signature of retryable function
+type RetryableFunc func() error
+
+// Default timer is a wrapper around time.After
+type timerImpl struct{}
+
+func (t *timerImpl) After(d time.Duration) <-chan time.Time {
+	return time.After(d)
+}
+
+func Do(retryableFunc RetryableFunc, opts ...Option) error {
+	var n uint
+
+	// default
+	config := newDefaultRetryConfig()
+
+	// apply opts
+	for _, opt := range opts {
+		opt(config)
+	}
+
+	if err := config.context.Err(); err != nil {
+		return err
+	}
+
+	// Setting attempts to 0 means we'll retry until we succeed
+	if config.attempts == 0 {
+		for err := retryableFunc(); err != nil; err = retryableFunc() {
+			if !IsRecoverable(err) {
+				return err
+			}
+
+			if !config.retryIf(err) {
+				return err
+			}
+
+			n++
+			config.onRetry(n, err)
+			select {
+			case <-config.timer.After(delay(config, n, err)):
```

Comment:
not sure if just me - but personally don't like a case statement with no body, since hard to see if intentional or not

Reply:
Feel free to send Avast a PR ;)